src/models/v3/operation.ts

```diff
@@ -0,0 +1,136 @@
+import type { v3 } from '../../spec-types';
+import { tilde } from '../../utils';
+import { BaseModel } from '../base';
+import { OperationAction, OperationInterface } from '../operation';
+import { ChannelInterface } from '../channel';
+import { Channels, ChannelsInterface } from '../channels';
+import { Channel } from './channel';
+import { SecurityRequirements } from '../security-requirements';
+import { SecurityRequirement } from './security-requirement';
+import { SecurityScheme } from './security-scheme';
+import { TagsInterface } from '../tags';
+import { bindings, extensions, tags } from './mixins';
+import { BindingsInterface } from '../bindings';
+import { ExtensionsInterface } from '../extensions';
+import { ServersInterface, Servers } from '../servers';
+import { ServerInterface } from '../server';
+import { MessagesInterface, Messages } from '../messages';
+import { MessageInterface } from '../message';
+import { ExternalDocumentationInterface } from '../external-docs';
+import { ExternalDocumentation } from './external-docs';
+import { OperationTraits, OperationTraitsInterface } from '../operation-traits';
+import { OperationTrait } from './operation-trait';
+
+export class Operation extends BaseModel<v3.OperationObject> implements OperationInterface {
+  id(): string {
+    return this._json['x-parser-id'];
+  }
+
+  hasId(): boolean {
+    return true;
+  }
+
+  action(): OperationAction {
+    return this._json.action;
+  }
+
+  isSend(): boolean {
+    return this._json.action === 'send';
+  }
+
+  isReceive(): boolean {
+    return this._json.action === 'receive';
+  }
+
+  servers(): ServersInterface {
+    const servers: ServerInterface[] = [];
+    const serversData: v3.ServerObject[] = [];
+    this.channels().forEach(channel => {
+      channel.servers().forEach(server => {
+        if (!serversData.includes(server.json())) {
+          serversData.push(server.json());
+          servers.push(server);
+        }
+      });
+    });
+    return new Servers(servers);
+  }
+  
+  messages(): MessagesInterface {
+    const messages: MessageInterface[] = [];
+    this.channels().forEach(channel => {
+      channel.messages().forEach(message => {
+        messages.push(message);
+      });
+    });
+    return new Messages(messages);
+  }
+
+  channels(): ChannelsInterface {
+    const channels: ChannelInterface[] = [];
+    const pointer = this._json.channel.address ? tilde(`/channels/${this._json.channel.address}`) : undefined;
+    channels.push(
+      this.createModel(Channel, this._json.channel, { id: this._json.channel['x-parser-id'], address: this._json.channel.address, pointer })
+    );
+    return new Channels(channels);
+  }
```

Comment:
Why are you doing it this way? Since we define with `$ref` channel in the operation, we only need to pass it to the collection without additional operations. Only problem is to retrieve proper pointer:

```suggestion
  channels(): ChannelsInterface {
    const channels = this._meta?.asyncapi?.parsed?.channels || {};
    let pointer: string = '';
    let channelId: string = '';
    for (const [channelName, channel] of Object.entries(channels)) {
      if (channel === this._json.channel) {
        channelId = channelName;
        pointer = `/channels/${channelName}`;
      }
    }
    const channel = this.createModel(Channel, this._json.channel, { id: channelId, pointer })
    return new Channels(channel);
  }
```

If there is anything you don't understand then let me know :)

Or I don't understand something.

Reply:
Did you try this implementation? I have a couple of concerns with this:

1. You're only checking for existence in `this._meta?.asyncapi?.parsed?.channels` but the reality is that a `$ref` can be pointing anywhere, even another file. Actually, most probably another file in many cases.
2. `if (channel === this._json.channel) {` is this strict equality gonna work? With `===` the two have to be the same object instance in memory, right? Or am I missing something?

Reply:
Ok, I pushed changes to the PR in latest commit - I know that there are quite a lot of changes, but we had a lot of missed methods or models didn't implement correctly interfaces. I will write unit tests and fix sonarcloud issues, don't worry. If you wanna, you can check implementation and ask if you have questions.